### PR TITLE
Implement REST endpoints for modifying voice state in stage channels. 

### PR DESCRIPTION
**Pull request by Neuheit**
## Description

Implements the REST endpoints documented for [Modify Current User Voice State](https://discord.com/developers/docs/resources/guild#modify-current-user-voice-state) and [Modify User Voice State](https://discord.com/developers/docs/resources/guild#modify-user-voice-state)

These endpoints also require the target user to be connected to a voice channel before they can be used, so I am not sure whether we should implement a join channel method via the gateway first (although they can always implement that themselves for now).

## Checklist

[//]: # "Put an x in [ ] to check the checkbox, e.g. [x]"

- [x] I discussed this PR with the maintainer(s) prior to opening it.
- [x] I read the [contributing guidelines](./.github/CONTRIBUTING.md).
- [x] I tested the changes in this PR.

### CODE REVIEW

File: src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using Disqord.Serialization.Json;
+
+namespace Disqord.Rest.Api
+{
+    public class ModifyCurrentUserVoiceStateJsonRestRequestContent : ModifyUserVoiceStateJsonRestRequestContent
+    {
+        [JsonProperty("request_to_speak_timestamp", NullValueHandling.Ignore)]
```
Comment (Quahu) [inline src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs:8]: Why is the null value ignored?

File: src/Disqord.Rest.Api/Content/Json/ModifyUserVoiceStateJsonRestRequestContent.cs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+﻿using Disqord.Serialization.Json;
+
+namespace Disqord.Rest.Api
+{
+    public class ModifyUserVoiceStateJsonRestRequestContent : JsonModelRestRequestContent
+    {
+        [JsonProperty("channel_id")]
+        public Snowflake ChannelId;
+
+        [JsonProperty("suppress", NullValueHandling.Ignore)]
```
Comment (Quahu) [inline src/Disqord.Rest.Api/Content/Json/ModifyUserVoiceStateJsonRestRequestContent.cs:10]: The null value ignoring here does nothing.

File: src/Disqord.Rest.Api/Routing/Route.Static.cs [hunk 3]
```diff
@@ -203,6 +203,10 @@ public static class Guild
             public static readonly Route GetWelcomeScreen = Get("guilds/{0:guild_id}/welcome-screen");
 
             public static readonly Route ModifyWelcomeScreen = Patch("guilds/{0:guild_id}/welcome-screen");
+
+            public static readonly Route ModifyCurrentUserVoiceState = Patch("guilds/{0:guild_id}/voice-states/@me");
```
Comment (Quahu) [inline src/Disqord.Rest.Api/Routing/Route.Static.cs:207]: ModifyCurrentMemberVoiceState, the request content and API client methods should be renamed accordingly

File: src/Disqord.Rest.Api/Routing/Route.Static.cs [hunk 4]
```diff
@@ -203,6 +203,10 @@ public static class Guild
             public static readonly Route GetWelcomeScreen = Get("guilds/{0:guild_id}/welcome-screen");
 
             public static readonly Route ModifyWelcomeScreen = Patch("guilds/{0:guild_id}/welcome-screen");
+
+            public static readonly Route ModifyCurrentUserVoiceState = Patch("guilds/{0:guild_id}/voice-states/@me");
+
+            public static readonly Route ModifyUserVoiceState = Patch("guilds/{0:guild_id}/voice-states/{1:user_id}");
```
Comment (Quahu) [inline src/Disqord.Rest.Api/Routing/Route.Static.cs:209]: ModifyMemberVoiceState

File: src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentUserActionProperties.cs [hunk 5]
```diff
@@ -1,6 +1,6 @@
 ﻿using System.IO;
 
-namespace Disqord.Rest
+namespace Disqord
```
Comment (Quahu) [inline src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentUserActionProperties.cs:3]: Don't change this, the missing namespaces on the other classes should probably be changed at some point instead.

File: src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Disqord
```
Comment (Quahu) [inline src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs:7]: Disqord.Rest

File: src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public sealed class ModifyCurrentVoiceStateActionProperties
+    {
+        public Optional<bool> Suppress { internal get; set; }
```
Comment (Quahu) [inline src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs:11]: IsSuppressed

File: src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+        public Optional<DateTimeOffset?> RequestToSpeakTimestamp { internal get; set; }
```
Comment (Quahu) [inline src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs:13]: RequestedToSpeakAt

File: src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs [hunk 9]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Disqord
+{
+    public sealed class ModifyCurrentVoiceStateActionProperties
```
Comment (Quahu) [inline src/Disqord.Rest/ActionProperties/Modify/ModifyCurrentVoiceStateActionProperties.cs:9]: The class is missing an internal constructor.

File: src/Disqord.Rest/Extensions/RestClientExtensions.Guild.cs [hunk 10]
```diff
@@ -429,5 +429,31 @@ public static async Task<IGuildPreview> FetchPreviewAsync(this IRestClient clien
...
+            {
+                ChannelId = channelId,
+                Suppress = properties.Suppress,
+                RequestToSpeakTimestamp = properties.RequestToSpeakTimestamp
+            };
+
+            return client.ApiClient.ModifyCurrentUserVoiceStateAsync(guildId, content, options);
+        }
+
+        public static Task ModifyVoiceStateAsync(this IRestClient client, Snowflake guildId, Snowflake userId, Snowflake channelId, bool? suppress = null, IRestRequestOptions options = null)
```
Comment (Quahu) [inline src/Disqord.Rest/Extensions/RestClientExtensions.Guild.cs:448]: `ModifyMemberVoiceStateAsync`, the `userId` parameter should be `memberId`, and the method should be using action properties.
I propose `ModifyMemberVoiceStateActionProperties` having the `IsSuppressed` property and `ModifyCurrentMemberVoiceStateActionProperties` inheriting it and adding `RequestedToSpeakAt`.
Same for the guild method.

File: src/Disqord.Rest/Extensions/RestClientExtensions.Guild.cs [hunk 11]
```diff
@@ -429,5 +429,31 @@ public static async Task<IGuildPreview> FetchPreviewAsync(this IRestClient clien
             var model = await client.ApiClient.FetchGuildPreviewAsync(guildId, options).ConfigureAwait(false);
             return new TransientGuildPreview(client, model);
         }
+
+        public static Task ModifyCurrentVoiceStateAsync(this IRestClient client, Snowflake guildId, Snowflake channelId, Action<ModifyCurrentVoiceStateActionProperties> action, IRestRequestOptions options = null)
```
Comment (Quahu) [inline src/Disqord.Rest/Extensions/RestClientExtensions.Guild.cs:433]: `ModifyCurrentMemberVoiceStateAsync`
Same for the guild method.

**Review by Quahu — CHANGES_REQUESTED** [review]

File: src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs (hunk 1 above)
Comment (Neuheit) [reply src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs:8]: It's optional so it doesn't have to be included, so I have that there to reduce the payload size. I can still remove it if you want.

File: src/Disqord.Rest.Api/Content/Json/ModifyUserVoiceStateJsonRestRequestContent.cs (hunk 2 above)
Comment (Neuheit) [reply src/Disqord.Rest.Api/Content/Json/ModifyUserVoiceStateJsonRestRequestContent.cs:10]: See above.

File: src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs (hunk 1 above)
Comment (Quahu) [reply src/Disqord.Rest.Api/Content/Json/ModifyCurrentUserVoiceStateJsonRestRequestContent.cs]: That's what the `Optional<T>` is for. Remove it.

**Review by Quahu — APPROVED** [review]

**Pull request merged**